src/librustc_metadata/index.rs

```diff
@@ -64,24 +64,20 @@ impl<'tcx> LazySeq<Index> {
     /// DefIndex (if any).
     #[inline(never)]
     pub fn lookup(&self, bytes: &[u8], def_index: DefIndex) -> Option<Lazy<Entry<'tcx>>> {
-        let words = &bytes_to_words(&bytes[self.position..])[..self.len];
-
-        debug!("Index::lookup: index={:?} words.len={:?}",
+        debug!("Index::lookup: index={:?} len={:?}",
                def_index,
-               words.len());
+               self.len);
 
-        let positions = match def_index.address_space() {
-            DefIndexAddressSpace::Low => &words[1..],
+        let i = def_index.as_array_index() + match def_index.address_space() {
+            DefIndexAddressSpace::Low => 0,
             DefIndexAddressSpace::High => {
                 // This is a DefIndex in the higher range, so find out where
                 // that starts:
-                let lo_count = u32::from_le(words[0].get()) as usize;
-                &words[lo_count + 1 .. ]
+                read_le_u32(&bytes[self.position..]) as usize
             }
         };
 
-        let array_index = def_index.as_array_index();
-        let position = u32::from_le(positions[array_index].get());
+        let position = read_le_u32(&bytes[self.position + (1 + i) * 4..]);
```

Comment:
So, [looking at the LLVM IR](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=e35d109e4716f35105dbb6fb19f719d3), this kind of the indexing is the culprit, I think.
This slicing, plus the `b[..4]` one in `read_le_u32` / `write_le_u32`, result in two checks, and the indexing is done with a `shl %i, 2` then an `i8` GEPi, instead of a `i32` GEPi.

At least it's not reading the bytes one by one from memory, that's what I was *really* worried about.
I'll try to come up with some arrangement which optimizes better.